test/functional/wallet_avoid_mixing_output_types.py

```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test output type mixing during coin selection
+
+A wallet may have different type UTXOs to choose from during coin selection,
+where output type is one of the following:
+    - BECH32M
+    - BECH32
+    - P2SH-SEGWIT
+    - LEGACY
+
+This test verifies that mixing different output types is avoided unless
+absolutely necessary. Both wallets start with zero funds. Alice mines
+enough blocks to have spendable coinbase outputs. Alice sends 10
+random value payments to Bob, creating 5 bech32 UTXOs and 5 P2SH UTXOs,
+for a total of 30BTC in Bob's wallet.
+
+Bob then sends random valued payments back to Alice, some of which need
+unconfirmed change, and we verify that none of these payments contain mixed
+inputs. Finally, Bob sends the remainder of his funds, which requires mixing.
+
+The payment values are random, but chosen such that they sum up to a specified
+total. This ensures we are not relying on specific values for the UTXOs,
+but still know when to expect mixing due to the wallet being close to empty.
+
+"""
+
+import random
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.blocktools import COINBASE_MATURITY
+
+ADDRESS_TYPES = [
+    "bech32m",
+    "bech32",
+    "p2sh-segwit",
+    "legacy",
+]
+
+
+def is_bech32_address(node, addr):
+    """Check if an address contains a bech32 output."""
+    addr_info = node.getaddressinfo(addr)
+    if addr_info['iswitness']:
+        return (
+            not addr_info['isscript'] and
+            addr_info['iswitness'] and addr_info['witness_version'] == 0
+        )
+    return False
+
+
+def is_bech32m_address(node, addr):
+    """Check if an address contains a bech32m output."""
+    addr_info = node.getaddressinfo(addr)
+    if addr_info['iswitness']:
+        return (
+            addr_info['isscript'] and
+            addr_info['iswitness'] and addr_info['witness_version'] == 1
+        )
+    return False
+
+
+def is_p2sh_address(node, addr):
+    """Check if an address contains a P2SH output."""
+    addr_info = node.getaddressinfo(addr)
+    return (
+        addr_info['isscript'] and
+        not addr_info['iswitness']
+    )
+
+
+def is_legacy_address(node, addr):
+    """Check if an address contains a legacy output."""
+    addr_info = node.getaddressinfo(addr)
+    return (not addr_info['isscript'] and not addr_info['iswitness'])
```

Comment:
In 784b03888e70a4744e7b387576f874ac18339dbb "test: functional test for new coin selection logic "

Not sure if this is relevant, but I think this may be inconsistent with the categorization pointed out by @achow101 previously, which IIUC prescribes that non-segwit P2SH addresses should be labeled as "legacy".

Reply:
it is inconsistent in the general sense, but for the test it is correct in that we are only generating `p2sh-segwit` outputs. i'll update the comment in the function, tho, to avoid any confusion